### DISCUSSION TITLE
How to preserve preserve gradient flow through preprocessor (AudioPreprocessor)?

### DISCUSSION THREAD

**Discussion by belson17** [NVIDIA-NeMo/NeMo#14463] [General]
Hi all, I'm trying to use a Nemo Parakeet model and do an optimization on the *input audio* (not the model weights), and so I want the gradients of the input with respect to the loss preserved so I can use SGD. However, the gradient flow is lost through the preprocessor (maybe in other places too) during the forward step of the model. Here's a simple self-contained example that shows the problem. 

```
import torch
import math
from nemo.collections.asr.models import EncDecCTCModelBPE

def create_temp_sine_wave(duration_sec=1.0, sample_rate=16000, freq=440.0):
    t = torch.linspace(0, duration_sec, int(sample_rate*duration_sec))
    waveform = 0.1 * torch.sin(2 * math.pi * freq * t)
    waveform = waveform.unsqueeze(0)  # shape (1, time)
    return waveform, sample_rate

def main():
    model_name = "stt_en_squeezeformer_ctc_xsmall_ls"
    model = EncDecCTCModelBPE.from_pretrained(model_name=model_name)
    print(f"preprocessor class name = {model.preprocessor.__class__.__name__}")
    print(f"featurizer class name = {model.preprocessor.featurizer.__class__.__name__}")
    waveform, sr = create_temp_sine_wave()
    audio_length = torch.tensor([waveform.shape[-1]])
    waveform.requires_grad_()
    waveform.retain_grad()

    print("Waveform requires_grad before preprocessing?", waveform.requires_grad)
    processed_feat, processed_len = model.preprocessor(input_signal=waveform, length=audio_length)
    print("Processed features require_grad?", processed_feat.requires_grad)

    # Backprop a dummy loss to test grad flow
    dummy_loss = processed_feat.sum()
    dummy_loss.backward()
    print(f"{waveform.grad=}")

if __name__ == "__main__":
    main()
```

The output shows:
```
preprocessor class name = AudioToMelSpectrogramPreprocessor
featurizer class name = FilterbankFeatures
Waveform requires_grad before preprocessing? True
Processed features require_grad? False
```

I need the last output to be `True`. 

After reviewing the code on github, even creating a new FilterbankFeatures instance with argument `use_args`  set to True, and overwriting AudioToMelSpectrogramPreprocessor.featurizer to this new FilterbankFeatures does not solve the issue. See below for relevant parts of the code.
https://github.com/NVIDIA/NeMo/blob/310a862cac55f7698c5bc4fe6c080cf4d359fe89/nemo/collections/asr/modules/audio_preprocessing.py#L111
https://github.com/NVIDIA/NeMo/blob/310a862cac55f7698c5bc4fe6c080cf4d359fe89/nemo/collections/asr/parts/preprocessing/features.py#L246

What might be the cause and what else can I do? Thank you!
